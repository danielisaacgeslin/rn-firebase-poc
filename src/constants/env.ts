import { Platform, StatusBar } from 'react-native';
import { Header } from 'react-navigation-stack';
import config from '../config/config.json';

export const ENV = {
  ENVIRONMENT: config.ENV,
  IS_DEV: config.ENV === 'development',
  IS_TEST: config.ENV === 'test',
  IS_QA: config.ENV === 'qa',
  IS_PROD: config.ENV === 'stage' || config.ENV === 'production',
  API: {
    URL: config.API_URL,
    MAX_RETRIES: 3,
    RETRY_TIMEOUT: config.API_RETRY_TIMEOUT,
    ENTITY: {
      AUTH: 'auth',
      USER: 'users',
      TODO: 'todos'
    }
  },
  PAGINATION: {
    LIMIT: 40
  },
  CHECK_FOR_UPDATES_INTERVAL: config.CHECK_FOR_UPDATES_INTERVAL,
  STORAGE_TIMEOUT: config.STORAGE_TIMEOUT,
  STORAGE_KEY: {
    AUTH: 'auth-data'
  },
  PLATFORM: {
    IS_ANDROID: Platform.OS === 'android',
    IS_IOS: Platform.OS === 'ios'
  },
  KEYBOARD_VERTICAL_OFFSET: (StatusBar.currentHeight || 0) + (Header.HEIGHT || 0) + 25,
  FIREBASE: {
    API_KEY: config.API_KEY,
    AUTH_DOMAIN: config.AUTH_DOMAIN,
    DATABASE_URL: config.DATABASE_URL,
    PROJECT_ID: config.PROJECT_ID,
    STORAGE_BUCKET: config.STORAGE_BUCKET,
    MESSAGING_SENDER_ID: config.MESSAGING_SENDER_ID,
    APP_ID: config.APP_ID
  }
};
