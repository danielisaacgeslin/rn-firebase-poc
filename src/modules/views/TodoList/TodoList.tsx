import React, { memo, useMemo, useCallback, useEffect } from 'react';
import { View, FlatList, Text } from 'react-native';

import { ENV } from '../../../constants';
import { GeneralModel, TodoModel, UserModel } from '../../models';
import { IEntityMap } from '../../../types';
import Todo from './components/Todo';
import styles from './styles';

export interface ITodoListProps {
  currentUser: UserModel.IUser;
  todoMap: IEntityMap<TodoModel.ITodo>;
  fetchTodoList: (query: GeneralModel.IApiQuery) => void;
}

const TodoList = ({ currentUser, todoMap, fetchTodoList }: ITodoListProps) => {
  const todoList = useMemo(() => Object.values(todoMap).sort((a, b) => (a._id > b._id ? -1 : 1)), [todoMap]);

  const keyExtractor = useCallback((item: TodoModel.ITodo): string => item._id, []);
  const renderTodo = useCallback(({ item }: { item: TodoModel.ITodo }) => <Todo todo={item} />, []);

  useEffect(() => {
    fetchTodoList({ page: 1, limit: ENV.PAGINATION.LIMIT, q: {} });
  }, []);

  return (
    <View style={styles.todoContainer}>
      <View style={styles.todoBody}>
        <FlatList testID="todo-list" data={todoList} extraData={todoList} renderItem={renderTodo} keyExtractor={keyExtractor} />
      </View>
    </View>
  );
};

export default memo(TodoList);
