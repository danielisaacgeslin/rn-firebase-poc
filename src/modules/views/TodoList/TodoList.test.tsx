import React from 'react';
import { render, RenderAPI } from 'react-native-testing-library';

import TodoList, { ITodoListProps } from './TodoList';
import { ENV } from '../../../constants';
import { getUser_1 } from '../../../test/entities';
import { TodoModel } from '../../models';

describe('TodoList', () => {
  let props: ITodoListProps;
  let wrapper: RenderAPI;

  beforeEach(() => {
    props = {
      currentUser: getUser_1(),
      todoMap: {
        '1': { _id: '1', title: '1', description: '1', status: TodoModel.Status.PENDING },
        '3': { _id: '3', title: '3', description: '3', status: TodoModel.Status.PENDING }
      },
      fetchTodoList: jest.fn()
    };
    wrapper = render(<TodoList {...props} />);
  });

  it('should render', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
    expect(props.fetchTodoList).toBeCalledWith({ page: 1, limit: ENV.PAGINATION.LIMIT, q: {} });
  });
});
