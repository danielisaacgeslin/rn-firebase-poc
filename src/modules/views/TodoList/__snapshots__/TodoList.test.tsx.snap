// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F6F6F6",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "_id": "3",
            "description": "3",
            "status": "pending",
            "title": "3",
          },
          Object {
            "_id": "1",
            "description": "1",
            "status": "pending",
            "title": "1",
          },
        ]
      }
      disableVirtualization={false}
      extraData={
        Array [
          Object {
            "_id": "3",
            "description": "3",
            "status": "pending",
            "title": "3",
          },
          Object {
            "_id": "1",
            "description": "1",
            "status": "pending",
            "title": "1",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="todo-list"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#1745D5",
                  "fontSize": 19,
                }
              }
            >
              3
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 15,
                }
              }
            >
              3
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 13,
                }
              }
            >
              created at: 
              undefined
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#1745D5",
                  "fontSize": 19,
                }
              }
            >
              1
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 15,
                }
              }
            >
              1
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 13,
                }
              }
            >
              created at: 
              undefined
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
